src/apps/virtio_net/virtio_net_app.lua

```diff
@@ -0,0 +1,68 @@
+module(..., package.seeall)
+
+local basic_apps = require("apps.basic.basic_apps")
+local pcap       = require("apps.pcap.pcap")
+local virtio_dev = require("apps.virtio_net.virtio_net_dev")
+local main       = require("core.main")
+
+local virtio_net = {}
+virtio_net.__index = virtio_net
+
+function virtio_net:new(args)
+   return setmetatable({
+      device = assert(virtio_dev.VGdev:new(args)),
+   }, self)
+end
+
+function virtio_net:stop()
+   self.device:close()
+end
+
+function virtio_net:push()
+   local dev = self.device
+   local l = self.input.rx
+   if not dev or not l then return end
+
+   while not link.empty(l) and dev:can_transmit() do
+      dev:transmit(link.receive(l))
+   end
+   dev:sync_transmit()
+end
+
+function virtio_net:pull()
+   local dev = self.device
+   local l = self.output.tx
+   if not dev or not l then return end
+
+   dev:sync_receive()
```

Comment:
on a pure polling driver, there seems to be no need for `:sync_receive()`; it's still there mostly as a placeholder for some notification suppression schemes that might help significantly with performance.  so far, none have yielded any consistent advantage, unfortunately.